frappe/core/doctype/submission_queue/submission_queue.py

```diff
@@ -0,0 +1,118 @@
+from frappe.desk.doctype.notification_log.notification_log import enqueue_create_notification
+from frappe.model.document import Document
+from frappe.utils import now
+from rq.exceptions import NoSuchJobError
+from rq.job import Job
+from frappe.utils.background_jobs import get_redis_conn
+
+
+class SubmissionQueue(Document):
+	@property
+	def created_at(self):
+		return self.creation
+
+	def insert(self, to_be_queued_doc: Document, action: str):
+		self.to_be_queued_doc = to_be_queued_doc
+		self.action_for_queuing = action
+		super().insert()
+
+	def lock(self):
+		self.to_be_queued_doc.lock()
+
+	def unlock(self):
+		# NOTE: this is called in execute_action method of Document class
+		# where get_doc is called hence we lose the to_be_queued_doc attribute
+		frappe.get_doc(self.ref_doctype, self.ref_docname).unlock()
+
+	def after_insert(self):
+		job = self.queue_action(
+			"queue_in_background",
+			to_be_queued_doc=self.to_be_queued_doc,
+			action_for_queuing=self.action_for_queuing,
+			timeout=600,
+		)
+		frappe.db.set_value(
+			self.doctype,
+			self.name,
+			{"job_id": job.id},
+			update_modified=False,
+		)
+
+	def queue_in_background(self, to_be_queued_doc: Document, action_for_queuing: str):
+		_action = action_for_queuing.lower()
+
+		if _action == "update":
+			_action = "submit"
+
+		try:
+			getattr(to_be_queued_doc, _action)()
+			values = {"status": "Completed"}
+		except Exception:
+			values = {"status": "Failed", "message": frappe.get_traceback()}
+			frappe.db.rollback()
+
+		values["ended_at"] = now()
+		frappe.db.set_value(self.doctype, self.name, values, update_modified=False)
+		self.notify(values["status"], action_for_queuing)
+
+	def notify(self, submission_status: str, action: str):
+		if submission_status == "Failed":
+			doctype = "Submission Queue"
+			docname = self.name
+			message = _("Submission of {0} {1} with action {2} failed")
+		else:
+			doctype = self.ref_doctype
+			docname = self.ref_docname
+			message = _("Submission of {0} {1} with action {2} completed successfully")
+
+		notification_doc = {
+			"type": "Alert",
+			"document_type": doctype,
+			"document_name": docname,
+			"subject": message.format(
+				frappe.bold(str(self.ref_doctype)), frappe.bold(self.ref_docname), frappe.bold(action)
+			),
+		}
+
+		notify_to = frappe.db.get_value("User", self.enqueued_by, fieldname="email")
+		enqueue_create_notification([notify_to], notification_doc)
+
+	@frappe.whitelist()
+	def unlock_doc(self):
+		if self.is_locked:
+			try:
+				Job(self.job_id, connection=get_redis_conn())
+				frappe.msgprint(_("Document already exists in queue!"))
+			except NoSuchJobError:
+				self.to_be_queued_doc.unlock()
+				self.status = "Failed"
+				self.save()
+				frappe.msgprint(_("Unlocked document as no such document exists in queue"))
+		else:
+			# failed, completed don't know at this point
+			self.status = "Failed"
+			self.save()
+			frappe.msgprint(_("Document is already unlocked"))
```

Comment:
```suggestion
		try:
			Job.fetch(self.job_id, connection=get_redis_conn())
			frappe.msgprint(_("Job for submission of the reference document exists in queue"), _("Failed to unlock"))
		except NoSuchJobError:
			self.unlock()
			frappe.msgprint(_("Reference Document Unlocked"))
```

TBH not really sure if the status can be considered `Failed` in the except block, since the job can also be killed/timed out after the whole submission scenario happens ..best to leave it in the `Queued` state?

Reply:
Plus i'm kinda unsure about the `Job.fetch` method - able to fetch the job even after it's completed

<img width="1185" alt="Screenshot 2022-10-08 at 3 22 40 AM" src="https://user-images.githubusercontent.com/32034600/194668046-6acada2c-4e2b-4387-a0c4-40aaa5f5b4cc.png">

maybe we can use this?

https://github.com/rq/rq/blob/4750eb43161cf6d2dee5cab4c20b7257476fbc1b/rq/queue.py#L215-L218

Reply:
`job.heart_beat` would be better

Reply:
We should also have a better way of enforcing job status for documents which are not in rq's queue but are marked as Queued